JS/main.js

```diff
@@ -0,0 +1,574 @@
+        drawingStates.push(canvas.toDataURL());
+    };
+};
+
+fillTool.addEventListener("click", fillCanvas);
+
+/* Undo function
+
+Defining the function which handles the undo event.
+
+We are updating the check checkStateNumber to be equal with the drawingStateNumber (which is used later at the redo button 
+to determin whether the user drew after the undo button in order to deactivate the redo button);
+
+We are updating the redoStates defined at the start of the code so that we can have a state to redo to, if needed;
+
+We are checking if there are any drawStates to which we can undo to;
+
+We are clearing the canvas before doing the undo, as the state previously saved would overlap the present canvas;
+
+We are "popping" the undo state from the drawingStates Array;
+
+We are importing the previous state (-1) as an img at 0,0 coordinates;
+
+Finally we have the two ways to call the event (one is present on the UI and the other one is the combination of CTRL + "z" for Windows or CMD + "z" for MacOS).
+*/
+
+function undo() {
+    checkStateNumber = drawingStateNumber;
+    redoStates.push(canvas.toDataURL());
+    if (drawingStates.length > 0) {
+        context.clearRect(0, 0, canvas.width, canvas.height);
+        drawingStates.pop();
+        const img = new Image();
+        img.src = drawingStates[drawingStates.length - 1];
+        img.addEventListener("load", () => {
+            context.drawImage(img, 0, 0);
+        });
+    };
+};
+
+window.addEventListener("keydown", (ev) => {
+    if (ev.key === "z" && ev.ctrlKey) {
+        undo();
+    };
+
+    if (ev.metaKey === "z" && ev.ctrlKey) {
+        undo();
+    };
+});
+
+undoButton.addEventListener("click", undo);
+
+/* Redo function
+
+Defining the function which handles the redo event.
+
+We "push" the actual drawing state to the drawingStates Array so that we have a state to undo to, if needed;
+
+We check if we have redoStates to Redo to (one gets added each time the Undo event is triggered);
+
+We are importing the previous state (-1) as an img at 0,0 coordinates, however this time around we clear the canvas inside the img;
+
+Finally, we are "popping" the latest Redo state from the redoStates Array.
+
+We have the two ways to call the event (one is present on the UI and the other one is the combination of CTRL + "x" for Windows or CMD + "x" for MacOS).
+
+*/
+
+function redo() {
+    drawingStates.push(canvas.toDataURL());
+    if (redoStates.length > 0) {
+        const img = new Image();
+        img.src = redoStates.pop();
+        img.addEventListener("load", () => {
+            context.clearRect(0, 0, canvas.width, canvas.height);
+            context.drawImage(img, 0, 0);
+        });
+    };
+};
+
+window.addEventListener("keydown", (ev) => {
+    if (drawingStateNumber > checkStateNumber) {
+        redoStates = [];
+    } else if (ev.key === "y" && ev.ctrlKey) {
+        redo();
+    } else if (ev.metaKey === "y" && ev.ctrlKey) {
```

Comment:
Meta înlocuiește Ctrl pe Mac, nu `key`.

```suggestion
    } else if (ev.key === "y" && ev.metaKey) {
```

Reply:
Din neatentie am pus in partea in care nu trebuia. Fixed.